src/sentry/tasks/on_demand_metrics.py

```diff
@@ -199,12 +238,64 @@ def process_widget_specs(widget_query_ids: list[int], *args, **kwargs) -> None:
     )
 
 
-def _set_widget_query_low_cardinality(
+def _get_widget_on_demand_specs(
+    widget_query: DashboardWidgetQuery,
+    organization: Organization,
+) -> Sequence[HashedMetricSpec]:
+    """
+    Saves on-demand state for a widget query.
+    """
+    # This can just be the first project we find, since spec hashes should not be project
+    # dependent. If spec hashes become project dependent then this may need to change.
+    project_for_query = Project.objects.filter(organization=organization).first()
+
+    if not project_for_query:
+        return []
+
+    widget_specs = convert_widget_query_to_metric(project_for_query, widget_query, True, False)
+
+    return widget_specs
+
+
+def _set_widget_on_demand_state(
+    widget_query: DashboardWidgetQuery,
+    specs: Sequence[HashedMetricSpec],
+    is_low_cardinality: bool | None,
+    enabled_features: Set[str],
+):
+    extraction_state = _determine_extraction_state(specs, is_low_cardinality, enabled_features)
+    spec_hashes = [hashed_spec[0] for hashed_spec in specs]
+
+    DashboardWidgetQueryOnDemand.objects.update_or_create(
+        dashboard_widget_query=widget_query,
+        defaults={
+            "spec_hashes": spec_hashes,
+            "extraction_state": extraction_state,
+        },
+    )
+
+
+def _determine_extraction_state(
```

Comment:
Where/when are we going to set `DISABLED_SPEC_LIMIT`?
Right now, we're okay since all of our organizations are below the limit but we will have to deal with it.
Perhaps we need to keep a counter for an org.
We can deal with this in a different PR. Just writing it as a task would do.